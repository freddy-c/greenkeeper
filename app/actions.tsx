"use server";

import { revalidatePath } from "next/cache";
import { z } from "zod";
import prisma from "./client";
import { Manufacturer } from "@prisma/client";

const db = prisma;

const schema = z.object({
    name: z.string().min(1).trim(),
});

export async function createManufacturer(
    data: z.infer<typeof schema>
): Promise<{
    success: boolean;
    manufacturer?: Manufacturer;
    issues?: {
        path: any;
        message: string;
    }[];
}> {
    const parsed = schema.safeParse(data);

    if (!parsed.success) {
        return {
            success: false,
            issues: parsed.error.issues.map((issue) => {
                return { path: issue.path, message: issue.message };
            }),
        };
    }

    const existingManufacturer = await db.manufacturer.findMany({
        where: {
            name: parsed.data.name,
        },
    });

    if (existingManufacturer.length > 0) {
        return {
            success: false,
            issues: [
                {
                    path: ["name"],
                    message: "This name is already in use",
                },
            ],
        };
    }

    try {
        const newManufacturer = await db.manufacturer.create({
            data: {
                name: parsed.data.name,
            },
        });

        revalidatePath("/");
        return {
            success: true,
            manufacturer: newManufacturer,
        };
    } catch (e) {
        return { success: false };
    }
}

export async function deleteManufacturer(
    prevState: {
        message: string;
    },
    formData: FormData
) {
    const schema = z.object({
        id: z.string().uuid(),
    });
    const parsed = schema.safeParse({
        id: formData.get("id"),
    });

    if (!parsed.success) {
        return { message: "Failed to delete manufacturer" };
    }

    const data = parsed.data;

    try {
        await db.manufacturer.delete({
            where: {
                id: data.id,
            },
        });

        revalidatePath("/");
        return { message: `Deleted manufacturer with id: ${data.id}` };
    } catch (e) {
        return { message: "Failed to delete manufacturer" };
    }
}
