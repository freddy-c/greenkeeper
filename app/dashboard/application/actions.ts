"use server";

import prisma from "@/app/client";
import { ApplicationSchema } from "./schemas";
import { revalidatePath } from "next/cache";

export async function createApplication(data: any): Promise<{
    success: boolean;
    message?: string;
    application?: any;
    issues?: {
        path: any;
        message: string;
    }[];
}> {
    const parsed = ApplicationSchema.safeParse(data);

    if (!parsed.success) {
        return {
            success: false,
            issues: parsed.error.issues.map((issue) => {
                return { path: issue.path, message: issue.message };
            }),
        };
    }

    try {
        // Extract the date from application.date and time from startTime and endTime
        const applicationDate = new Date(parsed.data.date);

        const startTime = new Date(parsed.data.startTime);
        const startDateTime = new Date(applicationDate);
        startDateTime.setHours(
            startTime.getHours(),
            startTime.getMinutes(),
            startTime.getSeconds()
        );

        const endTime = new Date(parsed.data.endTime);
        const endDateTime = new Date(applicationDate);
        endDateTime.setHours(
            endTime.getHours(),
            endTime.getMinutes(),
            endTime.getSeconds()
        );

        // Construct the data object for Prisma
        const applicationData = {
            date: applicationDate,
            sprayerId: parsed.data.sprayer?.value || null, // Handle optional sprayer
            operator: parsed.data.operator,
            method: parsed.data.method,
            startTime: startDateTime, // Combined date and time for start
            endTime: endDateTime, // Combined date and time for end
            waterVolume: parsed.data.waterVolume || null,
            temperature: parsed.data.temperature,
            windSpeed: parsed.data.windSpeed,
            observations: parsed.data.observations || "", // Optional observations
            ApplicationItems: {
                create: parsed.data.items.map((item: any) => ({
                    quantity: 0, // This can be calculated based on the rate/area logic
                    item: { connect: { id: item.itemId.value } }, // Add item connection
                })),
            },
            areasTreated: {
                connect: parsed.data.areas.map((area: any) => ({
                    id: area.managementZoneId.value,
                })),
            },
        };

        // Create the application in the database
        const newApplication = await prisma.application.create({
            data: applicationData,
        });

        revalidatePath("/application");

        return {
            success: true,
            application: newApplication,
        };
    } catch (e) {
        console.log(e);
        return { success: false, message: "Failed to create application" };
    }
}
