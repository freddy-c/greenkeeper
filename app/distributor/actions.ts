"use server";

import { z } from "zod";
import { AddDistributorSchema } from "./schemas";
import { Distributor } from "@prisma/client";
import prisma from "../client";
import { revalidatePath } from "next/cache";

const db = prisma;

export async function createDistributor(
    data: z.infer<typeof AddDistributorSchema>
): Promise<{
    success: boolean;
    distributor?: Distributor;
    issues?: {
        path: any;
        message: string;
    }[];
}> {
    // Validate input using Zod schema
    const parsed = AddDistributorSchema.safeParse(data);

    if (!parsed.success) {
        return {
            success: false,
            issues: parsed.error.issues.map((issue) => {
                return { path: issue.path, message: issue.message };
            }),
        };
    }

    // Check if the distributor already exists
    const existingDistributor = await db.distributor.findMany({
        where: {
            name: parsed.data.name,
        },
    });

    if (existingDistributor.length > 0) {
        return {
            success: false,
            issues: [
                {
                    path: ["name"],
                    message: "This name is already in use",
                },
            ],
        };
    }

    // Try creating the new distributor
    try {
        const newDistributor = await db.distributor.create({
            data: {
                name: parsed.data.name,
            },
        });

        revalidatePath("/"); // Revalidate the path to update UI after distributor creation
        return {
            success: true,
            distributor: newDistributor,
        };
    } catch (e) {
        return { success: false };
    }
}
