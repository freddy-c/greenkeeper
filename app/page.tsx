import { PrismaClient } from "@prisma/client";
import DeleteForm from "./delete-form";
import AddForm from "./add-form";

const db = new PrismaClient();

export default async function Home() {
    const manufacturers = await db.manufacturer.findMany();

    return (
        <div>
            <h1 className="text-lg font-semibold">Manufacturer List</h1>
            <ul>
                {manufacturers.map((manufacturer) => (
                    <li key={manufacturer.id} className="mb-4">
                        <h2>{manufacturer.name}</h2>
                        <DeleteForm manufacturerId={manufacturer.id} />
                    </li>
                ))}
            </ul>
            <AddForm />
        </div>
    );
}
