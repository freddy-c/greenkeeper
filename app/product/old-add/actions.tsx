"use server";

import { FormType, ProductType } from "@prisma/client";
import {
    FertilizerSchema,
    FormFinal,
    formSchema,
    GranularSchema,
    LiquidSchema,
    SolubleSchema,
} from "./schemas";
import prisma from "@/app/client";

export async function createProduct(data: FormFinal): Promise<{
    success: boolean;
    message?: string;
    product?: any;
}> {
    // Validate the data with Zod schema
    const parsed = formSchema.safeParse(data);

    if (!parsed.success) {
        return {
            success: false,
            message: parsed.error.errors.map((e) => e.message).join(", "),
        };
    }

    const { generalInfo, typeInfo, formInfo } = parsed.data;
    const { name, manufacturer, type, form } = generalInfo;
    const productType: ProductType = type as ProductType;
    const formType: FormType = form as FormType;

    try {
        // Create the product
        const product = await prisma.product.create({
            data: {
                name,
                manufacturerId: manufacturer.value,
                type: productType,
                form: formType,
            },
        });

        console.log(product);

        // Handle specific product type data
        switch (productType) {
            case "Fertilizer":
                const fertilizerData = FertilizerSchema.parse(typeInfo);
                await prisma.fertilizer.create({
                    data: {
                        productId: product.id,
                        ...fertilizerData,
                    },
                });
                break;

            // Add other product types if needed
        }

        // Handle form specific data
        switch (formType) {
            case "Liquid":
                const liquidData = LiquidSchema.parse(formInfo);
                await prisma.liquid.create({
                    data: {
                        productId: product.id,
                        specificGravity: liquidData.specificGravity,
                        waterRateMax: liquidData.maxWaterRate,
                        waterRateMin: liquidData.minWaterRate,
                    },
                });
                break;
            case "Granular":
                const granularData = GranularSchema.parse(formInfo);
                await prisma.granular.create({
                    data: {
                        productId: product.id,
                        granuleSizeMax: granularData.granuleSizeMax,
                        granuleSizeMin: granularData.granuleSizeMin,
                        minCuttingHeight: granularData.minCuttingHeight,
                        sgn: granularData.sgn,
                    },
                });
                break;
            case "Soluble":
                const solubleData = SolubleSchema.parse(formInfo);
                await prisma.soluble.create({
                    data: {
                        productId: product.id,
                        waterRateMax: solubleData.waterRateMax,
                        waterRateMin: solubleData.waterRateMin,
                    },
                });
                break;
        }

        return {
            success: true,
            product,
        };
    } catch (error) {
        console.log(error);
        return {
            success: false,
        };
    }
}
