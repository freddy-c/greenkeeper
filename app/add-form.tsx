"use client";

import { z } from "zod";
import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import { createManufacturer } from "@/app/actions";

const schema = z.object({
    name: z.string().min(1).trim(),
});

export default function AddForm() {
    const {
        register,
        setError,
        reset,
        handleSubmit,
        formState: { errors },
    } = useForm<z.infer<typeof schema>>({
        resolver: zodResolver(schema),
    });
    const onSubmit = async (data: z.infer<typeof schema>) => {
        const response = await createManufacturer(data);

        if (!response.success) {
            response.issues?.forEach((issue) => {
                issue.path.forEach((pathElement: any) => {
                    setError(pathElement, {
                        type: "manual",
                        message: issue.message,
                    });
                });
            });
        } else {
            reset();
        }
    };

    return (
        <form
            onSubmit={handleSubmit(onSubmit)}
            className="flex flex-col max-w-md space-y-4"
        >
            <div className="flex flex-col">
                <label htmlFor="name">Name</label>
                <input id="name" {...register("name")} className="border" />
                {errors.name && <p>{errors.name.message}</p>}
            </div>
            <input
                type="submit"
                value="Add Manufacturer"
                className="border hover:cursor-pointer"
            />
        </form>
    );
}
