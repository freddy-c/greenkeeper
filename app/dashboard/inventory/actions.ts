"use server";

import { revalidatePath } from "next/cache";
import { z } from "zod";
import prisma from "@/app/client";
import { AddItemSchema } from "@/app/schemas";

export async function editItem(
    itemId: string,
    data: z.infer<typeof AddItemSchema>
): Promise<{
    success: boolean;
    message?: string;
    item?: any;
    issues?: {
        path: any;
        message: string;
    }[];
}> {
    // Validate the incoming data with zod
    const parsed = AddItemSchema.safeParse(data);

    if (!parsed.success) {
        return {
            success: false,
            issues: parsed.error.issues.map((issue) => {
                return { path: issue.path, message: issue.message };
            }),
        };
    }

    try {
        // Construct the data object for Prisma
        const itemData = {
            productId: parsed.data.product.value, // Use the UUID value
            distributorId: parsed.data.distributor.value, // Use the UUID value
            price: parsed.data.price,
            purchaseDate: new Date(parsed.data.purchaseDate), // Assuming the date is a string in 'yyyy-MM-dd' format
            initialQuantity: parsed.data.initialQuantity,
            currentQuantity: parsed.data.currentQuantity,
        };

        // Update the item in the database
        const updatedItem = await prisma.item.update({
            where: { id: itemId },
            data: itemData,
        });

        // Optionally revalidate the cache/path if needed
        revalidatePath("/inventory");

        return {
            success: true,
            item: updatedItem,
        };
    } catch (e) {
        console.log(e);
        return { success: false, message: "Failed to update item" };
    }
}
