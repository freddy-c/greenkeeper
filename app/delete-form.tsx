"use client";

import { useFormState, useFormStatus } from "react-dom";
import { deleteManufacturer } from "./actions";

const initialState = {
    message: "",
};

function DeleteButton() {
    const { pending } = useFormStatus();

    return (
        <button
            type="submit"
            aria-disabled={pending}
            className="border hover:cursor-pointer"
        >
            Delete
        </button>
    );
}

export default function DeleteForm({
    manufacturerId,
}: {
    manufacturerId: string;
}) {
    const [state, formAction] = useFormState(deleteManufacturer, initialState);
    return (
        <form action={formAction}>
            <input type="hidden" name="id" value={manufacturerId} />
            <DeleteButton />
            {state.message}
        </form>
    );
}
