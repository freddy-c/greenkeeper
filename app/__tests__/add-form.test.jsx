import "@testing-library/jest-dom";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import AddForm from "../add-form";
import { createManufacturer } from "@/app/actions";

// Mock the createManufacturer function
jest.mock("../actions", () => ({
    createManufacturer: jest.fn(),
}));

describe("AddForm", () => {
    test("shows error message when name is empty", async () => {
        render(<AddForm />);
        fireEvent.click(
            screen.getByRole("button", { name: /add manufacturer/i })
        );

        await waitFor(() => {
            expect(
                screen.getByText(/string must contain at least 1 character/i)
            ).toBeInTheDocument();
        });
    });

    test("successfully submits form with valid name", async () => {
        createManufacturer.mockResolvedValue({ success: true });
        render(<AddForm />);
        const nameInput = screen.getByLabelText(/name/i);
        fireEvent.change(nameInput, { target: { value: "Test Manufacturer" } });
        fireEvent.click(
            screen.getByRole("button", { name: /add manufacturer/i })
        );

        await waitFor(() =>
            expect(createManufacturer).toHaveBeenCalledWith({
                name: "Test Manufacturer",
            })
        );
    });

    // Test for form reset after successful submission (optional)
    test("resets form after successful submission", async () => {
        createManufacturer.mockResolvedValue({ success: true });
        render(<AddForm />);
        const nameInput = screen.getByLabelText(/name/i);
        fireEvent.change(nameInput, { target: { value: "Test Manufacturer" } });
        fireEvent.click(
            screen.getByRole("button", { name: /add manufacturer/i })
        );

        await waitFor(() => expect(nameInput.value).toBe(""));
    });

    test("shows error message for duplicate name", async () => {
        // Mock subsequent API call to simulate existing manufacturer
        createManufacturer.mockResolvedValueOnce({
            success: false,
            issues: [
                { path: ["name"], message: "This name is already in use" },
            ],
        });

        render(<AddForm />);
        const nameInput = screen.getByLabelText(/name/i);
        fireEvent.change(nameInput, { target: { value: "Test Manufacturer" } });
        fireEvent.click(
            screen.getByRole("button", { name: /add manufacturer/i })
        );

        await waitFor(() => {
            expect(
                screen.getByText(/This name is already in use/i)
            ).toBeInTheDocument();
        });
    });
});
