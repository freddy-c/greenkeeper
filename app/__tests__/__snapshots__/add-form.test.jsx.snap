// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddForm renders correctly 1`] = `
<form
  className="flex flex-col max-w-md space-y-4"
  onSubmit={[Function]}
>
  <div
    className="flex flex-col"
  >
    <label
      htmlFor="name"
    >
      Name
    </label>
    <input
      className="border"
      id="name"
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
    />
  </div>
  <input
    className="border hover:cursor-pointer"
    type="submit"
    value="Add Manufacturer"
  />
</form>
`;
